kms-message/src/kms_endian_private.h

```diff
@@ -0,0 +1,165 @@
+#define KMS_UINT32_FROM_BE(v) KMS_UINT32_SWAP_LE_BE (v)
+#define KMS_UINT32_TO_BE(v) KMS_UINT32_SWAP_LE_BE (v)
+#define KMS_UINT64_FROM_LE(v) ((uint64_t) v)
+#define KMS_UINT64_TO_LE(v) ((uint64_t) v)
+#define KMS_UINT64_FROM_BE(v) KMS_UINT64_SWAP_LE_BE (v)
+#define KMS_UINT64_TO_BE(v) KMS_UINT64_SWAP_LE_BE (v)
+#elif defined(KMS_MESSAGE_BIG_ENDIAN)
+#define KMS_UINT16_FROM_LE(v) KMS_UINT16_SWAP_LE_BE (v)
+#define KMS_UINT16_TO_LE(v) KMS_UINT16_SWAP_LE_BE (v)
+#define KMS_UINT16_FROM_BE(v) ((uint16_t) v)
+#define KMS_UINT16_TO_BE(v) ((uint16_t) v)
+#define KMS_UINT32_FROM_LE(v) KMS_UINT32_SWAP_LE_BE (v)
+#define KMS_UINT32_TO_LE(v) KMS_UINT32_SWAP_LE_BE (v)
+#define KMS_UINT32_FROM_BE(v) ((uint32_t) v)
+#define KMS_UINT32_TO_BE(v) ((uint32_t) v)
+#define KMS_UINT64_FROM_LE(v) KMS_UINT64_SWAP_LE_BE (v)
+#define KMS_UINT64_TO_LE(v) KMS_UINT64_SWAP_LE_BE (v)
+#define KMS_UINT64_FROM_BE(v) ((uint64_t) v)
+#define KMS_UINT64_TO_BE(v) ((uint64_t) v)
+#else
+#error "The endianness of target architecture is unknown."
+#endif
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * __kms_uint16_swap_slow --
+ *
+ *       Fallback endianness conversion for 16-bit integers.
+ *
+ * Returns:
+ *       The endian swapped version.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static KMS_MSG_INLINE uint16_t
+__kms_uint16_swap_slow (uint16_t v) /* IN */
+{
+   return ((v & 0x00FF) << 8) | ((v & 0xFF00) >> 8);
+}
+
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * __kms_uint32_swap_slow --
+ *
+ *       Fallback endianness conversion for 32-bit integers.
+ *
+ * Returns:
+ *       The endian swapped version.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static KMS_MSG_INLINE uint32_t
+__kms_uint32_swap_slow (uint32_t v) /* IN */
+{
+   return ((v & 0x000000FFU) << 24) | ((v & 0x0000FF00U) << 8) |
+          ((v & 0x00FF0000U) >> 8) | ((v & 0xFF000000U) >> 24);
+}
+
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * __kms_uint64_swap_slow --
+ *
+ *       Fallback endianness conversion for 64-bit integers.
+ *
+ * Returns:
+ *       The endian swapped version.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static KMS_MSG_INLINE uint64_t
+__kms_uint64_swap_slow (uint64_t v) /* IN */
+{
+   return ((v & 0x00000000000000FFULL) << 56) |
+          ((v & 0x000000000000FF00ULL) << 40) |
+          ((v & 0x0000000000FF0000ULL) << 24) |
+          ((v & 0x00000000FF000000ULL) << 8) |
+          ((v & 0x000000FF00000000ULL) >> 8) |
+          ((v & 0x0000FF0000000000ULL) >> 24) |
+          ((v & 0x00FF000000000000ULL) >> 40) |
+          ((v & 0xFF00000000000000ULL) >> 56);
```

Comment:
Note: `long long` is a new C99 feature in the mongocrypt project (supported as an extension when C99 mode is not enabled).